---
title: "From vcf file to assembled file"
toc: true
number-sections: true
highlight-style: pygments
format:
  html: 
    code-fold: true
    html-math-method: katex
editor_options: 
  chunk_output_type: console
---

```{r}
#| label: load-pkgs
#| code-summary: "Packages"
#| message: false

pacman::p_load( tidyverse, fs, tictoc, processx)
```

You can add options to executable code like this

```{r}
#| echo: false

run(
  command = "zsh",
  args = c(
  "/home/liz30/project/PLEIOVAR/PLEIOVAR_rpkg/dev/sh/pull_variants_by_gene_from_vcf_v2.sh", "~/project/PLEIOVAR/PLEIOVAR_rpkg/inst/extdata/vcf",  "~/project/PLEIOVAR/PLEIOVAR_rpkg/inst/extdata/gene",   "~/project/PLEIOVAR/PLEIOVAR_rpkg/tmpout", "1>~/project/PLEIOVAR/PLEIOVAR_rpkg/tmpout/pull_by_gene.log", "2>~/project/PLEIOVAR/PLEIOVAR_rpkg/tmpout/vcftools_output.log"
  )
)
```

```{r}
system2("ls", "-la")
```

Get the paths after PLEIOVAR package installation

```{r}

shell_script <- path_package("PLEIOVAR", "shellscript", "pull_variants_by_gene_from_vcf_v2.sh")

vcf_dir <- path_package("PLEIOVAR", "extdata", "vcf")

gene_dir <- path_package("PLEIOVAR", "extdata", "gene")

```

Pull variants by gene from vcf files

```{r}

system2(
  command = "zsh",
  args = c(
  # "~/project/PLEIOVAR/PLEIOVAR_rpkg/dev/sh/pull_variants_by_gene_from_vcf_v2.sh", 
    pull_variants_script,
  #"~/project/PLEIOVAR/PLEIOVAR_rpkg/inst/extdata/vcf",  
    vcf_dir,
  # "~/project/PLEIOVAR/PLEIOVAR_rpkg/inst/extdata/gene",   
    gene_dir,
  #"~/project/PLEIOVAR/PLEIOVAR_rpkg/tmpout", 
    "tmpout",
  #"1>~/project/PLEIOVAR/PLEIOVAR_rpkg/tmpout/pull_by_gene.log", 
    "1>tmpout/pull_by_gene.log", 
  #"2>~/project/PLEIOVAR/PLEIOVAR_rpkg/tmpout/vcftools_output.log"
    "2>tmpout/vcftools_output.log"
  )
)

```
