---
title: "Metabolic Disease"
output:
  html_document:
    df_print: paged
    number_sections: TRUE
    toc: TRUE
    # toc_float: TRUE
    code_folding: hide
    code_download: TRUE
editor_options: 
  chunk_output_type: console
---

```{r message=FALSE, warning=FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, error = FALSE, message = FALSE, fig.width = 10, fig.height = 10)
pacman::p_load(tidyverse, devtools, fs, pacman, tictoc, furrr)
options(DT.options = list(pageLength = 5, scrollX=TRUE))
```


```{r}
options()
options(readr.show_col_types = FALSE)
```


```{r}
use_r("strsplit1")
use_r("pc_snp")
use_r("vcf2assemble")
use_r("pc_snp_one_gene")
use_r("pc_snp_one_chrom")
```

```{r}
load_all()
exists("strsplit1", where = globalenv(), inherits = FALSE)
exists("strsplit1", where = globalenv(), inherits = TRUE)
```



```{r}
check()
```

```{r}
use_mit_license()
document()
?strsplit1
```
```{r}
install()
library(PLEIOVAR)
```

```{r}
use_testthat()
use_test("strsplit1")
test()
```


```{r}
use_package("stringr")
use_package("magrittr", "rlang", "dplyr", "stats")
use_package("readr")
use_package("utils")
usethis::use_pipe()
rename_files("strsplit1", "str_split_one")
```


```{r}
use_readme_rmd()
build_readme()
```


```{r}
Y[, c(TRUE, VarSnps >= VARCUT)]
sum(round(cumsum(lambda)/sum(lambda), 4) == round(OPFcum_s,4))
```

## data

```{r}
x <- sample(1000)
usethis::use_data(x, mtcars)
```

```{r}
x <- sample(1000)
usethis::use_data(x, mtcars, internal = TRUE)
```
```{r}
use_data_raw("pc_snp")
```

Test for whole chroms.  

1. Genewise genotype

```{r}
library(PLEIOVAR)
# Get the dir path
vcf_dir <- "/data/test/vcf"
gene_dir <- "/data/test/gene"
out_dir <- "func_output/result_chrom22"
job_num <- 70
extension <- 50000
# Pull variants from vcf files
get_genewise_genotype(vcf_dir, gene_dir, out_dir, job_num, extension)
```

2. Assembled files

```{r}
tic("Multiple chroms")
assembled_dir <- "/home/liz30/project/PLEIOVAR/PLEIOVAR_rpkg/func_output/result"
plan(multisession, workers = 70)
dir_ls(assembled_dir, regexp = ".*chrom_[1-9].*") |>
path("vcftools_output") |>
    walk(function(chrom) {
      tic(chrom)
      unite_vcftools_output_one_chrom(chrom)
      toc()
     })

plan(sequential)
toc()
```


3. pc_snp


```{r}
# Process multiple chroms
# Set up workers. But use function walk(), not future_walk(), to avoid double paralleling at both chrom and gene levels.
tic("Multiple chroms")
gene_bed_dir <-  "/data/test/gene"
out_dir <-  "func_output/result"
plan(multisession, workers = 70)
dir_ls(gene_bed_dir, regexp = ".*chrom_[1-9].*") |>
  walk(function(gene_bed_file) {
    tic(gene_bed_file)
    pc_snp_one_chrom(gene_bed_file, out_dir)
    toc()
  })
plan(sequential)
toc()
```



