


#' Unite the three output files of vcftools for one gene
#'
#' Three files are generated by vcftools for each gene, gene.012 containing individual x SNP matrix, gene.012.indv containing
#' individual IDs in one column, gene.012.pos containing chrom and SNP coordinate in two columns. This function combines the
#' three files and save it as a rds file under "rd" folder, which will be created if not pre-exists, at the same level of
#' assembled dir, i.e, if assembled dir is chrom_20/vcf_output, the united rds file goes to chrom_20/rds. Only genes containing
#' non-zero variants are kept.
#'
#' @param vcftools_output  Dir path storing assembled files generated by vcftools
#' @param gene character, the name of a single gene
#'
#' The assembled dir path should follow the format chrom_i/vcf_output. The three files of a gene will be located according to the
#' naming rule mentioned above. The united file is saved as rds at chrom_i/rds and create this dir if it has not existed. The genes
#' with zero SNPs are not converted to rds.
#'
#' @return The status of write_rds function
#' @export
#'
#' @examples
#' # The rds file is saved at "../../assembled/variants_by_gene/chrom_tst/rds"
#'
#' d <- "result/chrom_20/vcftools_output"
#' unite_vcftools_output_one_gene(
#'     vcftools_output = d,
#'     gene = "gene_20_2"
#'   )
#'
#' # Gene "MIR155" has no SNPs, so no rds file saved
#' unite_vcftools_output_one_gene(
#'     vcftools_output = d,
#'     gene = "MIR155"
#'   )
#'
#'

unite_vcftools_output_one_gene <- function(
  vcftools_output = "result/chrom_20/vcftools_output",
  gene = "gene_20_2"
) {
  # Get variant position, i.e., variant matrix
  posi <-
    readr::read_tsv(
      file = fs::path(vcftools_output, gene, ext = "012.pos"),
      col_names = c("chrom", "posi"),
      col_types = readr::cols(
        .default = readr::col_integer()
      )
    )

  if (nrow(posi) > 0) { # The gene has > 0 SNPs
    # Get individual ID
    indv <-
      readr::read_tsv(
        file = fs::path(vcftools_output, gene, ext = "012.indv"),
        col_names = c("id"),
        col_types = readr::cols(
          .default = readr::col_character()
        )
      )
    # Read in variant data with column name as variant position
    # Attach individual ID
    d <-
      readr::read_tsv(
        file = fs::path(vcftools_output, gene, ext = "012"),
        col_names = c("order", as.character(posi$posi)),
        col_select = -1, # Remove the first "order" column
        col_types = readr::cols(
          .default = readr::col_integer()
        )
      ) |>
      dplyr::mutate( # Add inidividual id as the first column
        id = indv$id,
        .before = 1
      )

    # Save as rds
    rds_out_dir <-
      vcftools_output |>
      # Leave the last dir level out, i.e, vcf_output
      fs::path_dir() |>
      # Add one dir level, making "rds" reside alongside vcf_output
      fs::path("assembled_file") |>
      fs::dir_create()
    # if (!fs::dir_exists(rds_out_dir)) fs::dir_create(rds_out_dir)
    readr::write_rds(d, file = fs::path(rds_out_dir, gene, ext = "rds"))
  }
}



